src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Utility.cs

```diff
@@ -2212,6 +2150,80 @@ internal static unsafe nuint WidenAsciiToUtf16(byte* pAsciiBuffer, char* pUtf16B
             goto Finish;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static unsafe void WidenAsciiToUtf1_Vector<TVectorByte, TVectorUShort>(byte* pAsciiBuffer, char* pUtf16Buffer, ref nuint currentOffset, nuint elementCount)
+            where TVectorByte : unmanaged, ISimdVector<TVectorByte, byte>
+            where TVectorUShort : unmanaged, ISimdVector<TVectorUShort, ushort>
+        {
+            ushort* pCurrentWriteAddress = (ushort*)pUtf16Buffer;
+            // Calculating the destination address outside the loop results in significant
+            // perf wins vs. relying on the JIT to fold memory addressing logic into the
+            // write instructions. See: https://github.com/dotnet/runtime/issues/33002
+            nuint finalOffsetWhereCanRunLoop = elementCount - (nuint)TVectorByte.Count;
+            TVectorByte asciiVector = TVectorByte.Load(pAsciiBuffer + currentOffset);
+            if (!HasMatch<TVectorByte>(asciiVector))
+            {
+                (TVectorUShort utf16LowVector, TVectorUShort utf16HighVector) = Widen<TVectorByte, TVectorUShort>(asciiVector);
+                utf16LowVector.Store(pCurrentWriteAddress);
+                utf16HighVector.Store(pCurrentWriteAddress + TVectorUShort.Count);
+                pCurrentWriteAddress += (nuint)(TVectorUShort.Count * 2);
+                if (((int)pCurrentWriteAddress & 1) == 0)
+                {
+                    // Bump write buffer up to the next aligned boundary
+                    pCurrentWriteAddress = (ushort*)((nuint)pCurrentWriteAddress & ~(nuint)(TVectorUShort.Alignment - 1));
+                    nuint numBytesWritten = (nuint)pCurrentWriteAddress - (nuint)pUtf16Buffer;
+                    currentOffset += (nuint)numBytesWritten / 2;
+                }
+                else
+                {
+                    // If input isn't char aligned, we won't be able to align it to a Vector
+                    currentOffset += (nuint)TVectorByte.Count;
+                }
+                while (currentOffset <= finalOffsetWhereCanRunLoop)
+                {
+                    asciiVector = TVectorByte.Load(pAsciiBuffer + currentOffset);
+                    if (HasMatch<TVectorByte>(asciiVector))
+                    {
+                        break;
+                    }
+                    (utf16LowVector, utf16HighVector) = Widen<TVectorByte, TVectorUShort>(asciiVector);
+                    utf16LowVector.StoreAligned(pCurrentWriteAddress);
+                    utf16HighVector.StoreAligned(pCurrentWriteAddress + TVectorUShort.Count);
```

Comment:
This won't be correct if `canAlign` was false (the `else` path above was hit).

Reply:
You mean I cannot use 'StoreAligned' in that case and will have to use regular 'store'? Something like

```
(utf16LowVector, utf16HighVector) = Widen<TVectorByte, TVectorUShort>(asciiVector);
if (canAlign) {
       utf16LowVector.StoreAligned(pCurrentWriteAddress);
       utf16HighVector.StoreAligned(pCurrentWriteAddress + TVectorUShort.Count);
} else {
       utf16LowVector.Store(pCurrentWriteAddress);
       utf16HighVector.Store(pCurrentWriteAddress + TVectorUShort.Count);
}

currentOffset += (nuint)TVectorByte.Count;
pCurrentWriteAddress += (nuint)(TVectorUShort.Count * 2);
```

or I could move the check to the top and just have a non aligning implementation of `WidenAsciiToUtf1_Vector` which I can call rather than 2 branches

Reply:
The issue here is that you did:
```csharp
if (((int)pCurrentWriteAddress & 1) == 0)
{
    // Bump write buffer up to the next aligned boundary
    pCurrentWriteAddress = (ushort*)((nuint)pCurrentWriteAddress & ~(nuint)(TVectorUShort.Alignment - 1));
    nuint numBytesWritten = (nuint)pCurrentWriteAddress - (nuint)pUtf16Buffer;
    currentOffset += (nuint)numBytesWritten / 2;
}
else
{
    // If input isn't char aligned, we won't be able to align it to a Vector
    currentOffset += (nuint)TVectorByte.Count;
}
```

and so if the `else` path is hit, the data won't be actually aligned and the `StoreAligned` call would fault.


Just using `Store` is fine since on modern Intel processors (since Haswell), `movups` and `movaps` (same for `movdqu` vs `movdqa`, etc) have the same perf. The only case they differ is that `movups` (or `movdqu`, etc) for an actually unaligned address can incur a penalty if the load/store would cross a cache-line or page boundary. This penalty won't be possible if it's actually aligned (as would be the case for the main `if` path).

Reply:
Done